### PR TITLE
Presentation: Filtering APIs docs

### PR DESCRIPTION
**Pull request by saskliutas**
Added documentation for new filtering APIs and promoted them from `alpha` to `beta`. Added section to NextVersion.md about new features in Presentation library.

Need to wait for https://github.com/iTwin/itwinjs-core/pull/4954 and https://github.com/iTwin/itwinjs-core/pull/4957 to complete so this branch could be rebased and extract-api rerun.

Closes https://github.com/iTwin/itwinjs-backlog/issues/592
Closes https://github.com/iTwin/itwinjs-core/issues/4955

### CODE REVIEW

File: docs/changehistory/NextVersion.md [hunk 1]
```diff
@@ -95,6 +98,31 @@ Previously, iTwin.js used [DELETE](https://www.sqlite.org/pragma.html#pragma_jou
 - Failure to close a writeable briefcase may leave a "-wal" file. Previously, if a program crashed or exited with an open briefcase, it would leave the briefcase file as-of its last call to `IModelDb.saveChanges`. Now, there will be another file with the name of the briefcase with "-wal" appended. This is not a problem and the briefcase is completely intact, except that the briefcase file itself is not sufficient for copying (it will not include recent changes.) The "-wal" file will be used by future connections and will be deleted the next time the briefcase is successfully closed.
 - Attempting to copy an open-for-write briefcase file may not include recent changes. This scenario generally only arises for tests. If you wish to copy an open-for-write briefcase file, you must now call [IModelDb.performCheckpoint]($backend) first.
 
+## Presentation
+
+### Tree hierarchy levels filtering
```
Comment (grigasp) [inline docs/changehistory/NextVersion.md:103]:
```suggestion
### Hierarchy levels filtering
```

File: docs/changehistory/NextVersion.md [hunk 2]
```diff
@@ -17,6 +17,9 @@ Table of contents:
 - [Geometry](#geometry)
   - [Query mesh convexity](#query-mesh-convexity)
 - [Write-ahead logging](#write-ahead-logging)
+- [Presentation](#presentation)
+  - [Tree hierarchy levels filtering](#tree-hierarchy-levels-filtering)
```
Comment (grigasp) [inline docs/changehistory/NextVersion.md:21]:
```suggestion
  - [Hierarchy levels filtering](#hierarchy-levels-filtering)
```

File: docs/changehistory/NextVersion.md [hunk 3]
```diff
@@ -95,6 +98,31 @@ Previously, iTwin.js used [DELETE](https://www.sqlite.org/pragma.html#pragma_jou
...
 - Attempting to copy an open-for-write briefcase file may not include recent changes. This scenario generally only arises for tests. If you wish to copy an open-for-write briefcase file, you must now call [IModelDb.performCheckpoint]($backend) first.
 
+## Presentation
+
+### Tree hierarchy levels filtering
+
+Ability to filter individual hierarchy level in tree was added for the trees that uses [PresentationTreeDataProvider]($presentation-components). To enable this [PresentationTreeRenderer]($presentation-components) should be passed to [ControlledTree]($components-react) through [ControlledTreeProps.treeRenderer]($components-react):
```
Comment (grigasp) [inline docs/changehistory/NextVersion.md:105]:
```suggestion
Ability to filter individual hierarchy levels was added for tree components that use [PresentationTreeDataProvider]($presentation-components). To enable this, [PresentationTreeRenderer]($presentation-components) should be passed to [ControlledTree]($components-react) through [ControlledTreeProps.treeRenderer]($components-react):
```

File: docs/changehistory/NextVersion.md [hunk 4]
```diff
@@ -95,6 +98,31 @@ Previously, iTwin.js used [DELETE](https://www.sqlite.org/pragma.html#pragma_jou
...
+
+```ts
+return <ControlledTree
+  // other props
+  treeRenderer={(treeProps) => <PresentationTreeRenderer {...treeProps} imodel={props.imodel} modelSource={filteredModelSource} />}
```
Comment (grigasp) [inline docs/changehistory/NextVersion.md:110]: Props aren't very clear (e.g. what is `filteredModelSource`?). Could we just pass `imodel={imodel} modelSoure={modelSource}`?

File: docs/changehistory/NextVersion.md [hunk 5]
```diff
@@ -95,6 +98,31 @@ Previously, iTwin.js used [DELETE](https://www.sqlite.org/pragma.html#pragma_jou
...
+/>;
+```
+
+[PresentationTreeRenderer]($presentation-components) renders nodes with actions buttons for applying and clearing filters. Based on ruleset used in the tree some hierarchy level might not be filterable. In that case actions buttons for those hierarchy levels are not rendered. If applied filter does not produce any nodes `There are no child nodes matching current filter` message is renderer in that hierarchy level.
```
Comment (grigasp) [inline docs/changehistory/NextVersion.md:114]:
```suggestion
[PresentationTreeRenderer]($presentation-components) renders nodes with action buttons for applying and clearing filters. Some hierarchy levels might not be filterable depending on the presentation rules used to build them. In that case, action buttons for those hierarchy levels are not rendered. If applied filter does not produce any nodes, `There are no child nodes matching current filter` message is rendered in that hierarchy level.
```

File: presentation/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterProperty.tsx [hunk 6]
```diff
@@ -6,31 +6,43 @@
...
+  propertyDescription: PropertyDescription;
+  /** Full name of property class. Format: <schema-name>:<class-name>. */
+  fullClassName: string;
+  /** Label of property category */
+  categoryLabel?: string;
 }
 
-/** @alpha */
+/**
+ * Component for rendering property in [FilterBuilder]($component-react) property selector. Property category and
```
Comment (grigasp) [inline presentation/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterProperty.tsx:29]:
```suggestion
 * Component for rendering property in [FilterBuilder]($components-react) property selector. Property category and
```

File: presentation/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterProperty.tsx [hunk 7]
```diff
@@ -6,31 +6,43 @@
...
-/** @alpha */
+/**
+ * Props for [[PresentationInstanceFilterProperty]] component.
+ * @beta
+ */
 export interface PresentationInstanceFilterPropertyProps {
-  instanceFilterPropertyInfo: InstanceFilterPropertyInfo;
+  /** Description of property. */
+  propertyDescription: PropertyDescription;
+  /** Full name of property class. Format: <schema-name>:<class-name>. */
```
Comment (grigasp) [inline presentation/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterProperty.tsx:22]:
```suggestion
  /** Full name of property class. Format: `<schema-name>:<class-name>`. */
```

File: presentation/components/src/presentation-components/tree/controlled/PresentationTreeNodeRenderer.tsx [hunk 8]
```diff
@@ -16,15 +16,19 @@ import { ButtonGroup, IconButton } from "@itwin/itwinui-react";
...
 export interface PresentationTreeNodeRendererProps extends TreeNodeRendererProps {
   onFilterClick: (node: PresentationTreeNodeItem) => void;
   onClearFilterClick: (node: PresentationTreeNodeItem) => void;
 }
 
 /**
- * @alpha
+ * Component for rendering tree nodes produced by [[PresentationTreeDataProvider]]. Additionally to the default
+ * tree node renderer it renders actions buttons for filterable hierarchy levels and also correctly renders
+ * nodes that carry info messages.
```
Comment (grigasp) [inline presentation/components/src/presentation-components/tree/controlled/PresentationTreeNodeRenderer.tsx:30]:
```suggestion
 * Component for rendering tree nodes produced by [[PresentationTreeDataProvider]]. Additionally to the default
 * tree node renderer, it renders action buttons for filterable hierarchy levels and also correctly renders
 * nodes that carry info messages.
```

File: presentation/components/src/presentation-components/tree/controlled/PresentationTreeRenderer.tsx [hunk 9]
```diff
@@ -16,15 +16,19 @@ import { useHierarchyLevelFiltering } from "./UseHierarchyLevelFiltering";
...
 export interface PresentationTreeRendererProps extends TreeRendererProps {
   imodel: IModelConnection;
   modelSource: TreeModelSource;
 }
 
 /**
- * @alpha
+ * Tree renderer component that enables hierarchy level filtering in trees using [[PresentationTreeDataProvider]].
+ * It renders tree nodes with actions buttons for setting and clearing filters. [[PresentationInstanceFilterBuilderDialog]] is opened
+ * for creating filters for hierarchy levels.
```
Comment (grigasp) [inline presentation/components/src/presentation-components/tree/controlled/PresentationTreeRenderer.tsx:30]:
```suggestion
 * Tree renderer component that enables hierarchy level filtering in trees using [[PresentationTreeDataProvider]].
 * It renders tree nodes with action buttons for setting and clearing filters. [[PresentationInstanceFilterBuilderDialog]] is opened
 * for creating filters for hierarchy levels.
```

**Review by grigasp — CHANGES_REQUESTED** [review]
- `filter-builder-dialog.jpg` - could we hide the duplicate "Spatial Category" class?
- `filtered-tree.jpg` - "Filter apply" should be "Filter applied"?

And just a question... I wonder if it's really helpful to us or anyone else to make all filter builder APIs public rather than internal. I doubt if anyone is going to use those small building blocks, so not much of making them public. However, by making them public we have to keep stable API, which makes it harder to make enhancements. What do you think?

**Comment by saskliutas** [comment]
> And just a question... I wonder if it's really helpful to us or anyone else to make all filter builder APIs public rather than internal. I doubt if anyone is going to use those small building blocks, so not much of making them public. However, by making them public we have to keep stable API, which makes it harder to make enhancements. What do you think?

I was debating this myself while documenting various context used in `PropertyFilterBuilder` component. I thought if we expose everything then we won't need to do small changes to adapt to various possible use cases. However, since it is a new component and there are no usages we could make everything internal except `PropertyFilterBuilder` components and expose other pieces later when that is required. So if you prefer for making all those pieces internal I will change this.

File: docs/changehistory/NextVersion.md (hunk 4 above)
Comment (saskliutas) [reply docs/changehistory/NextVersion.md]: Copied it straight from test-app did not notices it used those variables. Simplified.

**Comment by grigasp** [comment]
> > And just a question... I wonder if it's really helpful to us or anyone else to make all filter builder APIs public rather than internal. I doubt if anyone is going to use those small building blocks, so not much of making them public. However, by making them public we have to keep stable API, which makes it harder to make enhancements. What do you think?
> 
> I was debating this myself while documenting various context used in `PropertyFilterBuilder` component. I thought if we expose everything then we won't need to do small changes to adapt to various possible use cases. However, since it is a new component and there are no usages we could make everything internal except `PropertyFilterBuilder` components and expose other pieces later when that is required. So if you prefer for making all those pieces internal I will change this.

I think it makes sense to keep internals `@internal` until there's a need for them to be public. Will provide us flexibility. And we can always change to `@public` afterwards.

**Review by grigasp — APPROVED** [review]

**Review by raplemie — APPROVED** [review]
Didnt read all docs lines to catch typos, but overall looks good.

**Pull request merged**